didcomm/protocols/forward/forward.py

```diff
@@ -44,14 +101,52 @@ async def wrap_in_forward(
 
     :return: a top-level packed Forward message as JSON string
     """
-    return json.dumps(
-        ForwardMessage(
-            body=ForwardBody(next=""),
-            id="",
-            type="https://didcomm.org/routing/2.0/forward",
-        ).as_dict()
+
+    # TODO construct routing keys
+    # - resolve alternative ones for serviceEndpoint
+    # - prepedn to the provided list
+
+    service_endpoint = await find_did_service(
+        to, resolvers_config
     )
 
+    routing_key_ids = service_endpoint.routing_keys
+
+    if not routing_key_ids:
+        # XXX might need to raise an error fo the cases when
+        #     no forward envelops are rpesented
+        logger.debug("No routing keys found: skipping forward wrapping")
+        return None
+
+    pack_config = pack_config or PackEncryptedConfig()
+    pack_params = pack_params or PackEncryptedParameters()
+
+    # wrap forward msgs in reversed order so the message to final
+    # recipient 'to' will be the innermost one
+    for _to, _next in zip(
+        routing_key_ids[::-1], (routing_key_ids[1:] + [to])[::-1]
+    ):
+        fwd_attach = Attachment(
+            data=AttachmentDataJson(packed_msg)
+        )
+
+        fwd_msg = ForwardMessage(
+            id=pack_params.didcomm_id_generator,
+            body=ForwardBody(next=_next),
+            type=DIDCommMessageProtocolTypes.FORWARD.value,
+            attachments=[fwd_attach],
+        )
+
+        fwd_msg_encrypted = await find_keys_and_anoncrypt(
+            fwd_msg.as_dict(), _to, pack_config.enc_alg_anon, resolvers_config
+        )
+
+        packed_msg = fwd_msg_encrypted.msg
```

Comment:
Do we use `packed_msg` anywhere?

Reply:
yes, that is for the next loop's iterations